File: framework/core/cpp/lib/src/yijinjing/io/locator.cpp
/*****************************************************************************
 * Copyright [www.kungfu-trader.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *****************************************************************************/

//
// Created by Keren Dong on 2021/5/21.
//

#include <cstdlib>
#include <kungfu/common.h>
#include <kungfu/yijinjing/common.h>
#include <regex>

namespace kungfu::yijinjing::data {

namespace fs = std::filesystem;
namespace es = longfist::enums;

std::string get_home() {
  auto kf_home = std::getenv("KF_HOME");
  if (kf_home != nullptr) {
    return kf_home;
  }
#ifdef _WINDOWS
  auto appdata = std::getenv("APPDATA");
  auto root = fs::path(appdata);
#elif __APPLE__
  auto user_home = std::getenv("HOME");
  auto root = fs::path(user_home) / "Library" / "Application Support";
#elif __linux__
  auto user_home = std::getenv("HOME");
  auto root = fs::path(user_home) / ".config";
#endif // _WINDOWS
  return (root / "kungfu" / "home" / "runtime").string();
}

locator::locator() : root_(get_home()) {}

bool locator::has_env(const std::string &name) const { return std::getenv(name.c_str()) != nullptr; }

std::string locator::get_env(const std::string &name) const { return std::getenv(name.c_str()); }

std::string locator::layout_dir(const location_ptr &location, es::layout layout) const {
  auto dir = root_ /                                     //
             es::get_category_name(location->category) / //
             location->group /                           //
             location->name /                            //
             es::get_layout_name(layout) /               //
             es::get_mode_name(location->mode);
  if (not fs::exists(dir)) {
    fs::create_directories(dir);
  }
  return dir.string();
}

std::string locator::layout_file(const location_ptr &location, es::layout layout, const std::string &name) const {
  auto path = fs::path(layout_dir(location, layout)) / fmt::format("{}.{}", name, es::get_layout_name(layout));
  return path.string();
}

std::string locator::default_to_system_db(const location_ptr &location, const std::string &name) const {
  auto sqlite_layout = es::layout::SQLITE;
  auto db_file = layout_file(location, sqlite_layout, name);
  if (not fs::exists(db_file)) {
    auto system_db_file = root_ /                                     //
                          es::get_category_name(location->category) / //
                          location->group /                           //
                          location->name /                            //
                          es::get_layout_name(sqlite_layout) /        //
                          es::get_mode_name(location->mode);
    fs::copy(system_db_file, db_file);
  }
  return db_file;
}

std::vector<uint32_t> locator::list_page_id(const location_ptr &location, uint32_t dest_id) const {
  std::vector<uint32_t> result = {};
  auto dest_id_str = fmt::format("{:08x}", dest_id);
  auto dir = fs::path(layout_dir(location, es::layout::JOURNAL));
  for (auto &it : fs::recursive_directory_iterator(dir)) {
    auto basename = it.path().stem();
    if (it.is_regular_file() and it.path().extension() == ".journal" and basename.stem() == dest_id_str) {
      auto index = std::atoi(basename.extension().string().c_str() + 1);
      result.push_back(index);
    }
  }
  return result;
}

static constexpr auto g = [](const std::string &pattern) { return fmt::format("({})", pattern); };

std::vector<location_ptr> locator::list_locations(const std::string &category, const std::string &group,
                                                  const std::string &name, const std::string &mode) const {
  fs::path search_path = root_ / g(category) / g(group) / g(name) / "journal" / g(mode);
  std::smatch match;
  std::regex search_regex(search_path.string());
  std::vector<location_ptr> result = {};
  for (auto &it : fs::recursive_directory_iterator(root_)) {
    auto path = it.path().string();
    if (it.is_directory() and std::regex_match(path, match, search_regex)) {
      auto l = location::make_shared(es::get_mode_by_name(match[4].str()),     //
                                     es::get_category_by_name(match[1].str()), //
                                     match[2].str(),                           //
                                     match[3].str(),                           //
                                     std::make_shared<locator>(root_.string()));
      result.push_back(l);
    }
  }
  return result;
}

std::vector<uint32_t> locator::list_location_dest(const location_ptr &location) const {
  std::unordered_set<uint32_t> set = {};
  auto dir = fs::path(layout_dir(location, es::layout::JOURNAL));
  for (auto &it : fs::recursive_directory_iterator(dir)) {
    auto basename = it.path().stem();
    if (it.is_regular_file() and it.path().extension() == ".journal") {
      set.emplace(std::stoul(basename.stem(), nullptr, 16));
    }
  }
  return std::vector<uint32_t>(set.begin(), set.end());
}
} // namespace kungfu::yijinjing::data

File: framework/core/python/kungfu/yijinjing/locator.py
import errno
import glob
import kungfu
import os
import shutil

from kungfu.yijinjing import *

lf = kungfu.__bindings__.longfist
yjj = kungfu.__bindings__.yijinjing


class Locator(yjj.locator):
    def __init__(self, root):
        yjj.locator.__init__(self)
        self._root = root

    def has_env(self, name):
        return os.getenv(name) is not None

    def get_env(self, name):
        return os.getenv(name)

    def layout_dir(self, location, layout):
        mode = lf.enums.get_mode_name(location.mode)
        category = lf.enums.get_category_name(location.category)
        p = os.path.join(
            self._root,
            category,
            location.group,
            location.name,
            lf.enums.get_layout_name(layout),
            mode,
        )
        try:
            os.makedirs(p)
        except OSError as e:
            if e.errno != errno.EEXIST:
                raise
        return p

    def layout_file(self, location, layout, name):
        return os.path.join(
            self.layout_dir(location, layout),
            "{}.{}".format(name, lf.enums.get_layout_name(layout)),
        )

    def default_to_system_db(self, location, name):
        file = os.path.join(
            self.layout_dir(location, lf.enums.layout.SQLITE),
            "{}.{}".format(name, lf.enums.get_layout_name(lf.enums.layout.SQLITE)),
        )
        if os.path.exists(file):
            return file
        else:
            system_location = yjj.location(
                lf.enums.mode.LIVE, lf.enums.category.SYSTEM, "etc", "kungfu", self
            )
            system_file = os.path.join(
                self.layout_dir(system_location, lf.enums.layout.SQLITE),
                "{}.{}".format(name, lf.enums.get_layout_name(lf.enums.layout.SQLITE)),
            )
            shutil.copy(system_file, file)
            return file

    def list_page_id(self, location, dest_id):
        page_ids = []
        for journal in glob.glob(
            os.path.join(
                self.layout_dir(location, lf.enums.layout.JOURNAL),
                hex(dest_id)[2:] + ".*.journal",
            )
        ):
            match = JOURNAL_PAGE_PATTERN.match(journal[len(self._root) + 1 :])
            if match:
                page_id = match.group(6)
                page_ids.append(int(page_id))
        return page_ids

    def list_locations(self, category, group, name, mode):
        search_path = os.path.join(self._root, category, group, name, "journal", mode)
        locations = []
        for journal in glob.glob(search_path):
            match = JOURNAL_LOCATION_PATTERN.match(journal[len(self._root) + 1 :])
            if match:
                category = match.group(1)
                group = match.group(2)
                name = match.group(3)
                mode = match.group(4)
                location = yjj.location(
                    MODES[mode], CATEGORIES[category], group, name, self
                )
                locations.append(location)
        return locations

    def list_location_dest(self, location):
        search_path = os.path.join(
            self._root,
            lf.enums.get_category_name(location.category),
            location.group,
            location.name,
            "journal",
            lf.enums.get_mode_name(location.mode),
            "*.journal",
        )
        readers = {}
        for journal in glob.glob(search_path):
            match = JOURNAL_PAGE_PATTERN.match(journal[len(self._root) + 1 :])
            if match:
                dest = match.group(5)
                page_id = match.group(6)
                if dest in readers:
                    readers[int(dest, 16)].append(page_id)
                else:
                    readers[int(dest, 16)] = [page_id]
        return list(readers.keys())
